File: update_playlists.py
#!/usr/bin/env python3
"""
update_playlists.py - Select the next unplayed episode per show and write an
annotated URI queue file for station.liq to consume.

Concurrency model mirrors fetch_podcasts.py:
  - Shares the same exclusive lockfile (state/radio.lock). If the fetcher is
    still running, this run logs a skip and exits 0 instead of hitting a
    locked database.
  - Databases run in WAL mode with a busy timeout as a second safety net.

Selection: for each show, pick the earliest episode with played=0, write
playlists/<slug>.txt as a single annotated URI line, then mark it played.
Works identically for archived (local file_path) and live (enclosure_url)
shows because both store their episodes in the same table.
"""

import argparse
import fcntl
import json
import logging
import os
import sqlite3
import sys
from datetime import datetime
from pathlib import Path

ROOT = Path(__file__).resolve().parent
CONFIG_PATH = ROOT / "config.json"

STATE_DIR = None
SUBS_DB = None
PLAYED_DB = None
PODCASTS_DIR = None
LOGS_DIR = None
PLAYLISTS_DIR = None
LOCK_FILE = None

def load_config():
    with open(CONFIG_PATH) as f:
        return json.load(f)

def init_paths():
    global STATE_DIR, SUBS_DB, PLAYED_DB, PODCASTS_DIR, LOGS_DIR, PLAYLISTS_DIR, LOCK_FILE
    cfg = load_config()
    storage = Path(cfg["storage"]).expanduser().resolve()
    STATE_DIR = storage / "state"
    SUBS_DB = STATE_DIR / "subscriptions.db"
    PLAYED_DB = STATE_DIR / "played.db"
    PODCASTS_DIR = storage / "podcasts"
    LOGS_DIR = storage / "logs"
    PLAYLISTS_DIR = storage / "playlists"
    LOCK_FILE = STATE_DIR / "radio.lock"

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(message)s",
    handlers=[
        logging.StreamHandler(sys.stdout),
    ],
)
log = logging.getLogger("update_playlists")

def log_error(msg):
    log.error(msg)

def _setup_logging():
    fh = logging.FileHandler(LOGS_DIR / "update.log")
    fh.setFormatter(logging.Formatter("%(asctime)s [%(levelname)s] %(message)s"))
    log.addHandler(fh)

# ---------------------------------------------------------------------------
# Shared schema definitions (must match fetch_podcasts.py).
# ---------------------------------------------------------------------------
SHOWS_COLUMNS = {
    "slug":        "TEXT PRIMARY KEY",
    "guid":        "TEXT NOT NULL UNIQUE",
    "name":        "TEXT NOT NULL",
    "feed_url":    "TEXT NOT NULL UNIQUE",
    "source":      "TEXT DEFAULT 'manual'",
    "opml_import": "INTEGER DEFAULT 0",
    "archived":    "INTEGER DEFAULT 1",
    "created_at":  "TEXT DEFAULT (datetime('now'))",
}

EPISODES_COLUMNS = {
    "id":            "INTEGER PRIMARY KEY AUTOINCREMENT",
    "show_slug":     "TEXT NOT NULL",
    "guid":          "TEXT NOT NULL",
    "title":         "TEXT",
    "file_path":     "TEXT",
    "enclosure_url": "TEXT",
    "runlength":     "INTEGER",
    "played":        "INTEGER DEFAULT 0",
    "played_at":     "TEXT",
}

def _connect(db_path):
    conn = sqlite3.connect(str(db_path), timeout=5.0)
    conn.row_factory = sqlite3.Row
    conn.execute("PRAGMA journal_mode=WAL")
    conn.execute("PRAGMA busy_timeout=5000")
    return conn

def _ensure_columns(conn, table, columns):
    exists = conn.execute(
        "SELECT name FROM sqlite_master WHERE type='table' AND name=?", (table,)
    ).fetchone()
    if exists is None:
        defs = ",\n        ".join(f"{name} {spec}" for name, spec in columns.items())
        extra = ""
        if table == "episodes":
            extra = "\n        ,UNIQUE(show_slug, guid)"
        conn.execute(f"CREATE TABLE {table} (\n        {defs}{extra}\n    )")
    else:
        existing = {row[1] for row in conn.execute(f"PRAGMA table_info({table})")}
        for name, spec in columns.items():
            if name not in existing:
                col_default = spec.split("DEFAULT", 1)[1].strip() if "DEFAULT" in spec else "NULL"
                conn.execute(f"ALTER TABLE {table} ADD COLUMN {name} {col_default}")
    conn.commit()

def open_subs_db():
    conn = _connect(SUBS_DB)
    _ensure_columns(conn, "shows", SHOWS_COLUMNS)
    return conn

def open_played_db():
    conn = _connect(PLAYED_DB)
    _ensure_columns(conn, "episodes", EPISODES_COLUMNS)
    conn.execute(
        "CREATE INDEX IF NOT EXISTS idx_episodes_show_played ON episodes (show_slug, played)"
    )
    conn.commit()
    return conn

# ---------------------------------------------------------------------------
# Mutual exclusion (shared with fetch_podcasts.py via the same lockfile).
# ---------------------------------------------------------------------------
_lock_fd = None

def acquire_lock():
    global _lock_fd
    STATE_DIR.mkdir(parents=True, exist_ok=True)
    fd = os.open(str(LOCK_FILE), os.O_CREAT | os.O_RDWR, 0o644)
    try:
        fcntl.flock(fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
    except OSError:
        os.close(fd)
        return False
    os.ftruncate(fd, 0)
    os.write(fd, str(os.getpid()).encode())
    _lock_fd = fd
    return True

def release_lock():
    global _lock_fd
    if _lock_fd is not None:
        try:
            fcntl.flock(_lock_fd, fcntl.LOCK_UN)
            os.close(_lock_fd)
        finally:
            _lock_fd = None

# ---------------------------------------------------------------------------
# Core logic
# ---------------------------------------------------------------------------
def _annotate_uri(runlength, title, uri):
    """Build a liquidsoap annotated URI line. Values must be double-quoted
    and separated by commas; the whole annotation precedes a colon before
    the URI. Escapes embedded double quotes minimally."""
    def q(v):
        return '"' + str(v).replace('"', '\\"') + '"'
    ann = f"annotate:liq_runlength={q(runlength)},liq_title={q(title)}:"
    return ann + uri

def select_next_episode(slug, played_db):
    row = played_db.execute(
        "SELECT guid, title, file_path, enclosure_url, runlength "
        "FROM episodes WHERE show_slug=? AND played=0 ORDER BY id ASC LIMIT 1",
        (slug,),
    ).fetchone()
    return row

def mark_as_played(slug, guid, played_db):
    played_db.execute(
        "UPDATE episodes SET played=1, played_at=datetime('now') "
        "WHERE show_slug=? AND guid=?",
        (slug, guid),
    )
    played_db.commit()

def write_queue_file(slug, ep, archived):
    if archived:
        uri = ep["file_path"]
    else:
        uri = ep["enclosure_url"]
    if not uri:
        return None
    line = _annotate_uri(ep["runlength"], ep["title"], uri)
    out = PLAYLISTS_DIR / f"{slug}.txt"
    out.write_text(line + "\n")
    return out

def update_all():
    subs_db = open_subs_db()
    played_db = open_played_db()
    shows = subs_db.execute("SELECT slug, name, archived FROM shows ORDER BY name").fetchall()
    updated = 0
    for show in shows:
        slug = show["slug"]
        archived = show["archived"] == 1
        ep = select_next_episode(slug, played_db)
        if ep is None:
            continue
        out = write_queue_file(slug, ep, archived)
        if out is None:
            log.warning("No playable URI for %s (%s); skipping.", show["name"], slug)
            continue
        mark_as_played(slug, ep["guid"], played_db)
        updated += 1
        log.info("Queued %s: %s -> %s", slug, ep["title"], out.name)
    subs_db.close()
    played_db.close()
    log.info("=== Update complete: %d show(s) queued ===", updated)

def json_summary():
    subs_db = open_subs_db()
    played_db = open_played_db()
    shows = subs_db.execute("SELECT slug, name FROM shows ORDER BY name").fetchall()
    result = {}
    for show in shows:
        slug = show["slug"]
        counts = played_db.execute(
            "SELECT COUNT(*) AS total, SUM(CASE WHEN played=0 THEN 1 ELSE 0 END) AS unplayed "
            "FROM episodes WHERE show_slug=?",
            (slug,),
        ).fetchone()
        total = counts["total"] or 0
        unplayed = counts["unplayed"] or 0
        result[slug] = {"name": show["name"], "total": total, "unplayed": unplayed, "played": total - unplayed}
    subs_db.close()
    played_db.close()
    print(json.dumps(result, indent=2))

def main():
    parser = argparse.ArgumentParser(description="Playlist updater for radio automation")
    parser.add_argument("--json", action="store_true", help="Emit JSON summary and exit")
    args = parser.parse_args()

    init_paths()
    STATE_DIR.mkdir(parents=True, exist_ok=True)
    LOGS_DIR.mkdir(parents=True, exist_ok=True)
    PLAYLISTS_DIR.mkdir(parents=True, exist_ok=True)
    _setup_logging()

    # --json is read-only and doesn't need the write lock.
    needs_lock = not args.json

    if needs_lock and not acquire_lock():
        log.info("Another radio process holds the lock; skipping this run.")
        return

    try:
        if args.json:
            json_summary()
        else:
            update_all()
    finally:
        if needs_lock:
            release_lock()

if __name__ == "__main__":
    main()

File: update_playlists.rb
#!/usr/bin/env jruby
# frozen_string_literal: true
#
# update_playlists.rb - Select next unplayed episode per show and write queue files
#
# For each show in subscriptions.db:
#   1. Count total/unplayed episodes in played.db
#   2. If all played (and archive=0): reset all to played=0, pick first, mark played=1
#   3. If unplayed > 0: pick one at random, mark played=1
#   4. Write <storage>/queue/<slug>.txt with annotated URI or "SKIP"
#
# Usage:
#   ./update_playlists.rb [--json]

require "sequel"
require "jdbc/sqlite3"
require "digest/sha1"
require "fileutils"
require "json"

SCRIPT_DIR = File.expand_path(File.dirname(__FILE__))
CONFIG_PATH = File.join(SCRIPT_DIR, "config.json")

def load_config
  raw = File.read(CONFIG_PATH)
  cfg = JSON.parse(raw)
  storage = cfg["storage"].to_s.strip
  raise "Missing 'storage' in config.json" if storage.empty?
  {
    storage:        storage,
    icecast_port:   cfg.dig("icecast", "port").to_i,
    mount_point:    cfg.dig("icecast", "mount_point").to_s,
    source_user:    cfg.dig("icecast", "source_username").to_s,
    source_pass:    cfg.dig("icecast", "source_password").to_s,
    gpodder_host:   cfg.dig("gpodder", "host").to_s,
    gpodder_user:   cfg.dig("gpodder", "username").to_s,
    gpodder_pass:   cfg.dig("gpodder", "password").to_s,
    gpodder_device: cfg.dig("gpodder", "device_id").to_s,
    gpodder_enable: cfg.dig("gpodder", "enable") == true
  }
end

CFG       = load_config
STATE_DIR = File.join(CFG[:storage], "state")
QUEUE_DIR = File.join(CFG[:storage], "queue")
LOG_DIR   = File.join(CFG[:storage], "logs")
SUBS_DB   = File.join(STATE_DIR, "subscriptions.db")
PLAYED_DB = File.join(STATE_DIR, "played.db")

MEDIA_DIRS = %w[music podcasts jingles announcements]
MEDIA_DIRS.each do |d|
  FileUtils.mkdir_p(File.join(CFG[:storage], d))
end
FileUtils.mkdir_p(STATE_DIR)
FileUtils.mkdir_p(QUEUE_DIR)
FileUtils.mkdir_p(LOG_DIR)

LOG_FILE = File.join(LOG_DIR, "update_playlists.log")
$log_fh = File.open(LOG_FILE, "a+")

def log(level, msg)
  ts = Time.now.strftime("%Y-%m-%d %H:%M:%S")
  line = "[#{ts}] [#{level}] #{msg}"
  $stdout.puts(line)
  $log_fh.write(line + "\n")
  $log_fh.flush
rescue Exception => e
  # Log file may be unavailable; fall back to stdout only
  $stdout.puts("[WARN] Could not write log: #{e.message}")
end

def table_exists?(db, tbl)
  db[:sqlite_master].where(type: "table", name: tbl).count > 0
rescue Exception => e
  false
end

def ensure_schema(db_subs, db_eps)
  unless table_exists?(db_subs, :shows)
    db_subs.execute <<-SQL
      CREATE TABLE IF NOT EXISTS shows (
        guid TEXT PRIMARY KEY,
        slug TEXT UNIQUE NOT NULL,
        title TEXT NOT NULL,
        feed_url TEXT NOT NULL,
        description TEXT DEFAULT '',
        category TEXT DEFAULT '',
        audio_only INTEGER DEFAULT 1,
        archive INTEGER DEFAULT 0,
        created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
      )
    SQL
  end
  unless table_exists?(db_eps, :episodes)
    db_eps.execute <<-SQL
      CREATE TABLE IF NOT EXISTS episodes (
        guid TEXT PRIMARY KEY,
        show_guid TEXT NOT NULL,
        title TEXT NOT NULL,
        enclosure_url TEXT NOT NULL,
        enclosure_type TEXT DEFAULT '',
        duration_sec INTEGER DEFAULT 0,
        pub_date TEXT DEFAULT '',
        local_path TEXT DEFAULT '',
        played INTEGER DEFAULT 0,
        downloaded INTEGER DEFAULT 0,
        fetched_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
      )
    SQL
  end
end

$stdout.sync = true

begin
  db_subs = Sequel.connect("jdbc:sqlite:" + SUBS_DB)
  db_eps  = Sequel.connect("jdbc:sqlite:" + PLAYED_DB)
ensure_schema(db_subs, db_eps)

  json_output = ARGV.include?("--json")
  results = []

  shows = db_subs[:shows].all
  if shows.empty?
    log("INFO", "No shows registered.")
  else
    shows.each do |show|
      slug     = show[:slug]
      show_g   = show[:guid]
      archive  = show[:archive] || 0

      total    = db_eps[:episodes].where(show_guid: show_g).count
      unplayed = db_eps[:episodes].where(show_guid: show_g, played: 0).count

      chosen = nil

      if total == 0
        log("INFO", "#{slug}: no episodes yet, SKIP")
      elsif unplayed == 0 && archive == 0
        # Cycle: reset all to unplayed, pick first, mark played
        db_eps[:episodes].where(show_guid: show_g).update(played: 0)
        ep = db_eps[:episodes].where(show_guid: show_g).order(:pub_date.asc).first
        if ep
          db_eps[:episodes].where(guid: ep[:guid]).update(played: 1)
          chosen = ep
          log("INFO", "#{slug}: cycled, picked '#{ep[:title]}'")
        end
      elsif unplayed > 0
        eps = db_eps[:episodes].where(show_guid: show_g, played: 0).all
        ep  = eps.sample
        db_eps[:episodes].where(guid: ep[:guid]).update(played: 1)
        chosen = ep
        log("INFO", "#{slug}: picked '#{ep[:title]}' (#{unplayed} unplayed)")
      end

      qf = File.join(QUEUE_DIR, "#{slug}.txt")
      if chosen
        dur   = chosen[:duration_sec].to_i
        annot = "annotate:liq_runlength=\"#{dur}\",liq_title=\"#{chosen[:title]}\""
        uri   = chosen[:enclosure_url]
        File.write(qf, "#{annot}:#{uri}\n")
      else
        File.write(qf, "SKIP\n")
      end

      results << {
        "slug"     => slug,
        "total"    => total,
        "unplayed" => unplayed,
        "picked"   => chosen ? chosen[:title] : nil
      }
    end
  end

  if json_output
    puts JSON.pretty_generate(results)
  end

  log("INFO", "Done. Processed #{results.size} shows.")

  begin
    db_subs.disconnect
  rescue Exception => e
    log("WARN", "Error disconnecting subs: #{e.message}")
  end
  begin
    db_eps.disconnect
  rescue Exception => e
    log("WARN", "Error disconnecting eps: #{e.message}")
  end
  $log_fh.close

rescue Interrupt
  log("INFO", "Interrupted.")
  exit 1
rescue Exception => e
  log("ERROR", "#{e.class}: #{e.message}")
  log("ERROR", e.backtrace.first(5).join("\n"))
  exit 1
end

main if __FILE__ == $PROGRAM_NAME

